core/server/worker/src/main/java/alluxio/worker/netty/AbstractWriteHandler.java

```diff
@@ -119,23 +119,29 @@ public void channelRead(ChannelHandlerContext ctx, Object object) throws Excepti
 
     RPCProtoMessage msg = (RPCProtoMessage) object;
     Protocol.WriteRequest writeRequest = msg.getMessage().asWriteRequest();
-    // Only initialize (open the writers) if this is the first packet in the block/file.
-    if (writeRequest.getOffset() == 0) {
 
-      // Expected state: context equals null as this handler is new for request, or the previous
-      // context is not active (done / cancel / abort). Otherwise, notify the client an illegal
-      // state. Note that, we reset the context before validation msg as validation may require to
-      // update error in context.
-      try (LockResource lr = new LockResource(mLock)) {
-        Preconditions.checkState(mContext == null || !mContext.isPacketWriterActive());
+    try (LockResource lr = new LockResource(mLock)) {
+      boolean isContextNull = mContext == null;
+      if (isContextNull) {
+        // When mContext is null, create a new one as catching exceptions and replying errors
+        // leverages data structures in context, regardless of the request is valid or not.
+        // TODO(binfan): remove the dependency on an instantiated request context which is required
+        // to reply errors to client side.
         mContext = createRequestContext(writeRequest);
       }
-    }
+      // Only initialize (open the writers) if this is the first packet in the block/file.
+      if (writeRequest.getOffset() == 0) {
+        // Expected state: context equals null as this handler is new for request, or the previous
+        // context is not active (done / cancel / abort). Otherwise, notify the client an illegal
+        // state. Note that, we reset the context before validation msg as validation may require to
+        // update error in context.
+        Preconditions.checkState(isContextNull || !mContext.isPacketWriterActive());
```

Comment:
this should be `mContext.isDone`? I don't think `!isPacketWriterActive` is an accurate boolean for this state.

Also, could you rename `isContextNull` to something more accurate (since it is no longer true here even though the value is true)

Reply:
fixed. ptal